realm/realm-library/src/objectServer/java/io/realm/internal/async/RealmEventStreamTaskImpl.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2020 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal.async;
+
+import java.io.IOException;
+import java.util.concurrent.locks.ReentrantLock;
+
+import io.realm.internal.Util;
+import io.realm.internal.objectserver.EventStream;
+import io.realm.mongodb.App;
+import io.realm.mongodb.AppException;
+import io.realm.mongodb.ErrorCode;
+import io.realm.mongodb.RealmEventStreamTask;
+import io.realm.mongodb.mongo.events.BaseChangeEvent;
+
+public class RealmEventStreamTaskImpl<T> implements RealmEventStreamTask<T> {
+    private final String name;
+    private final Executor<T> executor;
+    private volatile EventStream<T> eventStream;
+    private volatile boolean isCancelled;
+    private final ReentrantLock lock;
+
+    public RealmEventStreamTaskImpl(final String name, final Executor<T> executor) {
+        Util.checkNull(executor, "name");
+        Util.checkNull(executor, "executor");
+
+        this.lock = new ReentrantLock();
+        this.executor = executor;
+        this.name = name;
+    }
+
+    synchronized EventStream<T> getEventStream() throws IOException {
+        if (eventStream == null) {
+            eventStream = executor.run();
+        }
+
+        return this.eventStream;
+    }
+
+    @Override
+    public BaseChangeEvent<T> getNext() throws IOException {
+        if (lock.tryLock()) {
```

Comment:
With this implementation, you can still change between Async/Sync execution. I don't know if this was intended?

E.g ... this would work:

```
watcher.getNext()
watcher.getAsync()
```

If intended, it should probably be named `getNextAsync()`? Alternatively, we set a mode boolean when calling either `getNext/getAsync()` and throw if the wrong get method is called. So you choose a mode when calling the first get method. Thoughts?